Finish gps support

tblaha opened the pull request:
To be able to generate identical output to bb-log-viewer also if GPS is included, I did the following:
- added a `last_gps_home_frame` variable to the `ctx` that is updated everytime GPS_HOME is parsed
- implement the predictors for GPS coordinates (latitude and longitude separately, which is caught explicitely in `reader.py`)
- ensured that only intra, slow and GPS fields are added to field_names (GPS home isn't written out by bb-log-viewer)
- add GPS data to the frame the same way slow frames have been
- updated `parser_test.py`. I ran 2 separate test cases through it, one with GPS and one without. However, more testing would be beneficial. @atomgomba would you still have some test `.csv` lying around?

atomgomba committed:
Bump version to 0.3.0

atomgomba committed:
Initial work on GPS support

atomgomba committed:
update spelling dictionary

tblaha committed:
finish GPS implementation

atomgomba commented:
@tblaha Once again, thank you for your contribution, I really appreciate it! :) Please rebase this onto master

tblaha commented:
No problem!
I use this tool to read data into a pandas dataframe and do the fixed-point to floating-point unit conversions. Would you be interested to add this, for instance as a `.to_pandas(convert_units=True)` method, or something similar?

atomgomba commented:
@tblaha Sorry for the delays, I have limited time, but I'm planning to test this today. If all goes well I'm going to push changes to the documentation, including README, and finally create a new PyPi release :)

atomgomba commented on orangebox/predictors.py:
```diff
@@ -64,6 +64,18 @@ def _increment(new: Number, ctx: Context) -> Number:
 
+@map_to(7, predictor_map)
+def _home_coord_0(new: Number, ctx: Context) -> Number:
+    # TODO: store GPS frame history
```
This TODO comment is now obsolete, please remove it

atomgomba commented on orangebox/predictors.py:
```diff
@@ -64,6 +64,18 @@ def _increment(new: Number, ctx: Context) -> Number:
+@map_to(7, predictor_map)
+def _home_coord_0(new: Number, ctx: Context) -> Number:
+    # TODO: store GPS frame history
+    return ctx.last_gps_home_frame.data[0] + new if ctx.last_gps_home_frame.data else 0
```
Please put the second member of this addition in parentheses for readability and extract this long expression into a variable named `result`

atomgomba replied:
This returns an `int`, please divide `result` by `10_000_000` so it becomes directly usable by consumers

atomgomba commented on orangebox/predictors.py:
```diff
@@ -64,6 +64,18 @@ def _increment(new: Number, ctx: Context) -> Number:
+
+@map_to(256, predictor_map)
+def _home_coord_1(new: Number, ctx: Context) -> Number:
+    # TODO: store GPS frame history
```
Obsolete comment, please remove

atomgomba commented on orangebox/predictors.py:
```diff
@@ -64,6 +64,18 @@ def _increment(new: Number, ctx: Context) -> Number:
+@map_to(256, predictor_map)
+def _home_coord_1(new: Number, ctx: Context) -> Number:
+    # TODO: store GPS frame history
+    return ctx.last_gps_home_frame.data[1] + new if ctx.last_gps_home_frame.data else 0
```
Please extract this long expression as well and add parentheses

atomgomba replied:
Also divide the result by `10_000_000`

atomgomba commented on orangebox/reader.py:
```diff
@@ -140,6 +140,9 @@ def _build_field_defs(self):
                     field_defs[frame_type] = [FieldDef(frame_type) for _ in range(len(header_value))]
                 prop = header_key.split(" ", 2)[-1]
                 for i, framedef_value in enumerate(header_value):
+                    if (field_defs[frame_type][i].name == "GPS_coord[1]") \
```
Please remove the parentheses, it's not required in Python

atomgomba replied:
Please extract the expression `field_defs[frame_type][i].name` into a variable called `fdef_name` and do not wrap the condition

atomgomba commented on orangebox/parser.py:
```diff
@@ -62,9 +62,13 @@ def set_log_index(self, index: int):
-        for fdef in reader.field_defs.values():
-            self._field_names += filter(lambda x: x is not None and x not in self._field_names,
-                                        map(lambda x: x.name, fdef))
+        for ftype in [FrameType.INTRA, FrameType.SLOW, FrameType.GPS]:
```
Nice catch! :+1:

atomgomba commented on orangebox/parser.py:
```diff
@@ -143,9 +161,27 @@ def frames(self) -> Iterator[Frame]:
-                frame = Frame(ftype, frame.data + last_slow.data)
+
+            # add in extra frames (GPS, GPS_HOME and SLOW)
+            edata = []
```
Please give a more descriptive name to this variable, maybe `extra_data`?

atomgomba commented on parser_test.py:
```diff
@@ -13,14 +13,31 @@ def compare(frame_index, parser, frame, csv_frame, show_all_fields):
-            fdef = reader.field_defs[FrameType.SLOW][i - fdeflen]
+
+        fdef = None
+        if (i < intra_inter_len):
```
No parentheses please

atomgomba commented on parser_test.py:
```diff
@@ -13,14 +13,31 @@ def compare(frame_index, parser, frame, csv_frame, show_all_fields):
+        if (i < intra_inter_len):
+            fdef = reader.field_defs[FrameType.INTER][i]
+        else:
+            curMaxLen = intra_inter_len
```
Please use `snake_case` naming (everywhere in this file)

atomgomba reviewed (changes requested):
The decoded output looks correct, I have made some remarks on the code

atomgomba commented:
I just realized just blindly dividing by `10_000_000` would mess with the altitude value, if my thinking is right. Feel free to disregard this part, we can fix this once this is merged in a separate PR

atomgomba committed:
Remove ignored folder, fix conflict

tblaha committed:
improve code clarity

tblaha commented:
Thanks for the review, looks like I need to brush up on some PEP8! I addressed all your comments and ran the tests again.
I agree that processing variables (like converting units or converting fixed point to floating point) should probably be done in other methods. It's also not done by bb-log-viewer; it just dumps the raw values into the CSV for all other fields as well.

atomgomba commented:
@tblaha Good job, thank you! :partying_face:
